util/stump-nm/nm-api.lisp

```diff
@@ -0,0 +1,124 @@
+  ((interface :initform "org.freedesktop.NetworkManager.Settings")))
+
+(defclass setting-connection (dbus-object)
+  ((interface :initform "org.freedesktop.NetworkManager.Settings.Connection")))
+
+(defun get-settings (root)
+  (make-instance 'settings
+                 :path "/org/freedesktop/NetworkManager/Settings"
+                 :bus (bus root)))
+
+(defclass connection-settings ()
+  ((id :initarg :id :reader id)
+   (uuid :initarg :uuid :reader uuid)
+   (connection :initarg :connection :reader connection)
+   (type :initarg :type :reader cs-type)))
+
+(defun get-connection-settings (settings)
+  (mapcar (lambda (connection-path)
+            (block lambda
+              (let* ((connection (make-instance 'setting-connection
+                                                :path connection-path
+                                                :bus (bus settings)))
+                     (connection-settings (dbus-call connection "GetSettings")))
+                (dolist (setting connection-settings)
+                  (when (string= (first setting) "connection")
+                    (return-from lambda
+                      (let ((settings (second setting)))
+                        (make-instance
+                         'connection-settings
+                         :connection connection
+                         :id (first (assoc-value settings "id" :test #'string=))
+                         :uuid (first (assoc-value settings "uuid" :test #'string=))
+                         :type (first (assoc-value settings "type" :test #'string=))))))))))
+          (getprop settings "Connections")))
+
+(defclass wireless-device (dbus-object)
+  ((interface :initform "org.freedesktop.NetworkManager.Device.Wireless")))
+
+(defclass access-point (dbus-object)
+  ((interface :initform "org.freedesktop.NetworkManager.AccessPoint")
+   (active :initarg :active :reader active)
+   (ssid :accessor ssid)
+   (frequency :accessor frequency)
+   (strength :accessor strength)))
+
+(defun device-access-points (device)
+  (with-introspected-object (object (bus device) (path device) *service*)
+    (let ((active-access-point (getprop device "ActiveAccessPoint")))
+      (mapcar (lambda (access-point-path)
+                (prog1-let (ap (make-instance 'access-point
+                                              :path access-point-path
+                                              :bus (bus device)
+                                              :active (string= access-point-path
+                                                               active-access-point)))
+                  (setf (ssid ap) (access-point-ssid ap)
+                        (frequency ap) (getprop ap "Frequency")
+                        (strength ap) (getprop ap "Strength"))))
+              (object (interface device) "GetAllAccessPoints")))))
+
+(defun access-point-ssid (ap)
+  (format nil "~{~a~}"
+          (mapcar #'code-char
+                  (getprop ap "Ssid"))))
```

Comment:
iiuc this code first transforms octets to chars and then collects them in a string, if so you can do the same by using `map` and specifying string as the result-type.
```suggestion
(defun access-point-ssid (ap)
  (map 'string #'code-char (getprop ap "Ssid")))
```

However this approach only works for code-points that fit in one 8-bit byte. ej.
```lisp
(let ((octets #(227 130 179 227 131 162 227 131 179 227 131 170 227 130 185 227 131 151 227 129 140 227 129 153 227 129 141 227 129 167 227 129 153)))
  (map 'string #'code-char octets)) ; => "ã³ã¢ã³ãªã¹ããããã§ã"
```

To remedy this please use the helper `stumpwm::utf8-to-string` which is a thin wrapper over `sb-ext:octets-to-string`.


```lisp
(let ((octets #(227 130 179 227 131 162 227 131 179 227 131 170 227 130 185 227 131 151 227 129 140 227 129 153 227 129 141 227 129 167 227 129 153)))
  (stumpwm::utf8-to-string octets)) ; => "コモンリスプがすきです"
```


sb-ext:octets-to-string or the wrapper we provide `utf8-to-string`

Reply:
Very good point! I've used babel instead, it's already loaded and OCTETS-TO-STRING is actually exported.